Planet-based acquisition option

AaronGullickson opened the pull request:
This feature adds a planet-based acquisition (PBA) in addition to the default acquisition method. It is designed to take account of where on the map the unit is and provide results to users that are simpler and more intuitive than the default method. 

## Basic Procedure

- Collect all planets within X*30 light years (specified by users up to X=5, default X=2) of the current location. Sort planets by a rough measure of the number of jumps and system transit time. All of this also includes the current planet which is searched first. Empty planets are also removed. 
- Loop through planets. For each part/unit type being searched for make an acquisition roll with the normal target number but also applying modifiers for the tech/industry/output ratings of the system (modifiers are customizable by user). If this succeeds, then the admin has "found a contact". The user may then roll normal acquisition rolls to acquire items, until they have acquired all of them or they fail a roll. Once they fail a roll, their source is tapped out and they must move to the next planet to search.
- Delivery times are equal to `(jumps-1)*7+(sending planet transit time)+(receiving planet transit time)+(jumps+1)d6 days`. The delivery time is the same for all units of the same type from a given planet. So you won't get 16 different Large Lasers delivered at different times, but shipments of multiple large lasers being delivered at the same time. 

Unless they have the verbose option turned on (not recommended), users won't see all of these details. They will just see when parts were found on a planet, how many, and when they will arrive. 

Because searches are done on multiple planets, this method is generally more likely to find parts than the default method, even with the added check, but this can be easily adjusted for by increasing the time between searches or adding general modifiers to make searches harder. Delivery times are also likely to be shorter, but that depends on the user's setting. 

One of the main advantages of this method (aside from ease of interpretation), is that it makes unit location on the map relevant for acquisitions. If you are in the core, say on Outreach, you are going to have a much easier time finding parts than if you are out in the periphery somewhere pirate hunting. It also integrates universe information, which is cool, although it will be cooler if we can impute all the missing values for socio-industrial level on most planets. Right now, it defaults to C if a planet has no known socio-industrial level. 

The user can also set whether they can be re-supplied from planets owned by certain factions based on the same faction relationships that are used to generate OpFor for AtB. Basically, you can choose all, neutral or ally, ally, or only your faction. 

## Code Information

Most of the action is taking place in a new function `Campaign.goShopping` that I re-factored from a couple of places in `ShoppingList` and `Campaign.acquireEquipment`. 

I wrote this with an eye to extendability. All of the planet based mods are in a function `Planet.getAcquisitionMods` which makes it easy to add in other information about planets as we develop that information (e.g. factories, supply planets).

AaronGullickson commented:
Ok, I will work on the style stuff. Thats a good point about getting clan equipment from the IS and vice versa. That would make sense in 3145 but not in 3050 although in 3145 I am guessing its still harder to acquire from the IS than the clans. I will add an additional modifier for clan equipment from IS as well as a button to disable clan equipment from the IS entirely. Just don't ask me to code in the Sea Fox aimags.

neoancient commented:
I thought about Sea Fox but chose not to mention it.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/Planet.java:
```diff
@@ -875,6 +878,105 @@ public double getDistanceTo(double x, double y) {
     }
 
+    //adapted from interstellar map panel because OOP
+    public boolean isEmpty(DateTime when) {
```
added

AaronGullickson commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -1875,24 +1876,181 @@ public Person getLogisticsPerson() {
+	                	int totalQuantity = 0;
+	                	//use same delivery time for the batch of same part type on the same planet
+	                	int transitTime = calculatePartTransitTime(planet);
+	                	if(shoppingItem.getQuantity() > 0 
```
The first attempt is the "identify contact" attempt and feeds in a true boolean to ```initialAttempt``` in ```acquireEquipment``` while the remaining checks feed in a false boolean for that. So I don't think they can be combined. However it does make me realize that there is a bug where it is incrementing totalQuantity the first time and I could also simply the while statement below it.

AaronGullickson replied:
Actually, I think I should probably refactor the initial roll for a contact into a different method.  That should simplify the code and make it easier to follow.

neoancient replied:
I see. I didn't pay close enough attention to the arguments.

AaronGullickson commented:
Ok, I did a bit of refactoring in goShopping and acquireEquipment to simplify the code in c9be07b. There is now a campaign.findContactForAcquisition method that is called first and separately from the acquireEquipment method.

AaronGullickson commented on MekHQ/src/mekhq/campaign/market/ShoppingList.java:
```diff
@@ -71,7 +71,20 @@
         shoppingList = new ArrayList<IAcquisitionWork>();
     }
+    
+    public ShoppingList(ArrayList<IAcquisitionWork> items) {
```
This has been done.

AaronGullickson commented:
Ok, I believe I have added all of the requested changes. I have also added two new options. One option disallows getting clan parts from non-clan factions. The second applies a penalty to getting clan parts from non-clan factions, which is in addition to the general penalty for acquiring clan parts.

neoancient commented:
Good, but now there are mixed tabs and spaces in `Campaign`. I noticed them in `findContactForAcquisition`

AaronGullickson commented:
I think I have corrected all of the whitespace issues.
